lib/dispatch.go

```diff
@@ -0,0 +1,217 @@
+		obj := reflect.New(c.InType)
+		return obj.Interface()
+	}
+	return nil
+}
+
+// lookup finds the callable structure with the given method name
+func (r *RegMethodSet) lookup(method string) (*callable, bool) {
+	if c, ok := r.reg[method]; ok {
+		return &c, true
+	}
+	return nil, false
+}
+
+type callable struct {
+	Impl    interface{}
+	Func    reflect.Value
+	InType  reflect.Type
+	OutType reflect.Type
+}
+
+// RegisterMethods iterates the methods provided by the lib API, and makes them visible to dispatch
+func (inst *Instance) RegisterMethods() {
+	reg := make(map[string]callable)
+	inst.registerOne("fsi", &FSIImpl{}, reg)
+	inst.regMethods = &RegMethodSet{reg: reg}
+}
+
+func (inst *Instance) registerOne(ourName string, impl interface{}, reg map[string]callable) {
+	implType := reflect.TypeOf(impl)
+	// Iterate methods on the implementation, register those that have the right signature
+	num := implType.NumMethod()
+	for k := 0; k < num; k++ {
+		m := implType.Method(k)
+		lowerName := strings.ToLower(m.Name)
+		funcName := ourName + "." + lowerName
+
+		// Validate the parameters to the method
+		// should have 3 input parameters: (receiver, scope, input struct)
+		// should have 2 output parametres: (output value, error)
+		// TODO(dustmop): allow variadic returns: error only, cursor for pagination
+		f := m.Type
+		if f.NumIn() != 3 {
+			log.Errorf("%s: bad number of inputs: %d", funcName, f.NumIn())
+			continue
+		}
+		if f.NumOut() != 2 {
+			log.Errorf("%s: bad number of outputs: %d", funcName, f.NumOut())
+			continue
+		}
+		// First input must be the receiver
+		inType := f.In(0)
+		if inType != implType {
+			log.Errorf("%s: first input param should be impl, got %v", funcName, inType)
+			continue
+		}
+		// Second input must be a scope
+		inType = f.In(1)
+		if inType.Name() != "Scope" {
+			log.Errorf("%s: second input param should be scope, got %v", funcName, inType)
+			continue
+		}
+		// Third input is a pointer to the input struct
+		inType = f.In(2)
+		if inType.Kind() != reflect.Ptr {
+			log.Errorf("%s: third input param must be a struct pointer, got %v", funcName, inType)
+			continue
+		}
+		inType = inType.Elem()
+		if inType.Kind() != reflect.Struct {
+			log.Errorf("%s: third input param must be a struct pointer, got %v", funcName, inType)
+			continue
+		}
+		// First output is anything
+		outType := f.Out(0)
+		// Second output must be an error
+		outErrType := f.Out(1)
+		if outErrType.Name() != "error" {
+			log.Errorf("%s: second output param should be error, got %v", funcName, inType)
+			continue
+		}
+
+		// Save the method to the registration table
+		reg[funcName] = callable{
+			Impl:    impl,
+			Func:    m.Func,
+			InType:  inType,
+			OutType: outType,
+		}
+		log.Infof("%d: registered %s(*%s) %v", k, funcName, inType, outType)
+	}
+}
+
+// methodEndpoint returns a method name and returns the API endpoint for it
+func methodEndpoint(method string) APIEndpoint {
+	if method == "fsi.write" {
```

Comment:
Why is this hardcoded here?
Actually, I don't really understand the encoding schema here.
`fsi.write` is a special case as the route has the extra prefix. On the remaining routes we simply take the `package.method` fromat and return a `/method/` route, however that's not always the case. `registry`, `remote`, `profile`, `IPFS` are all in a similar boat as `fsi.write`. 
Additionally a lot of routes do not support a trailing `/` and will fail in that case.

I know we need a translation layer for `method name` > `endpoint` but I'd rather have it just be directly assigned instead of handling all the edge cases with auto parsing. Maybe move the parsing function into `api/api.go` and use that to set the `methodEndpoints` along with the handler so that in future extensions it's hard to miss.

It's a bit anti-pattern in that it surfaces some lib level knowledge up to API but given that we rely on that same HTTP API for our RPC layer, we have to be strict about it regardless and the Qri implementation for the API needs to be aware of those details. Other users can skip that part if they don't intend on using the HTTP RPC layer at all.

Reply:
I probably should have left a comment, but I definitely intended this hardcoded part to be temporary in there's a better plan. You're absolutely right that the correspondence between API paths and methods needs to be better codified, for now this PR is keeping it the same by pointing out the special case. We should make this do something better in a follow-up change.